File: frontend/js/views/spia.js
// SPIA — single-premium immediate annuity: monthly income from a lump sum,
// the payout rate, and total received, via /calc/spia. Updates live.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const FIELDS = [
    ['premium_usd', 'Premium (lump sum, $)', 100000],
    ['payout_years', 'Payout period (years)', 20],
    ['annual_rate_pct', 'Assumed rate (%)', 5],
];

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 });
const money0 = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 });
const VIEW = 'spia';
let lastReport = null;
let lastBody = null;

export async function renderSpia(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.spia.h1.title">// SPIA INCOME</span></h1>
        <p class="muted small" data-i18n="view.spia.hint.intro">
            A single-premium immediate annuity turns a lump sum into guaranteed income starting
            now. The monthly payment is the annuity that exhausts the premium over the payout
            period at the insurer's assumed rate — the same math as a loan payment, in reverse.
            The payout rate is the annual income as a percent of the premium. (Period-certain
            approximation; a true life annuity also prices in mortality.) Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.spia.h2.inputs">The annuity</h2>
            <form id="spia-form" class="inline-form">
                ${FIELDS.map(([key, label, def]) => `
                    <label><span data-i18n="view.spia.label.${key}">${label}</span>
                        <input type="number" step="0.01" min="0" name="${key}" value="${def}" required></label>
                `).join('')}
            </form>
            <div id="spia-tools" class="ce-toolbar"></div>
            <button type="button" id="spia-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="spia-sens" class="ce-sens"></div>
        </div>
        <div id="spia-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#spia-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        const body = {};
        for (const [key] of FIELDS) body[key] = Number(fd.get(key)) || 0;
        return body;
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcSpia(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || t('view.spia.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#spia-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'spia.csv' });
    mount.querySelector('#spia-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['monthly_income_usd', r.monthly_income_usd],
        ['annual_income_usd', r.annual_income_usd],
        ['payout_rate_pct', r.payout_rate_pct],
        ['total_received_usd', r.total_received_usd],
        ['interest_earned_usd', r.interest_earned_usd],
    ];
}

async function renderResult(mount, r, body, tok) {
    const el = mount.querySelector('#spia-result');
    // Line chart: monthly income as the assumed rate sweeps 0 -> 10%.
    const xs = enh.linspace(0, 10, 13);
    const pts = await Promise.all(xs.map(async (rate) => {
        const rr = await api.calcSpia({ ...body, annual_rate_pct: rate });
        return { x: rate, y: rr ? rr.monthly_income_usd : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'rate %', ylabel: 'monthly $' });
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.spia.h2.result">The income</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.spia.card.monthly">Monthly income</div>
                    <div class="value pos">${money(r.monthly_income_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.spia.card.annual">Annual income</div>
                    <div class="value">${money0(r.annual_income_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.spia.card.payout">Payout rate</div>
                    <div class="value">${Number(r.payout_rate_pct).toFixed(2)}%</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <thead><tr><th data-i18n="view.spia.col.line">Line</th><th data-i18n="view.spia.col.amount">Amount</th></tr></thead>
                <tbody>
                    <tr><td data-i18n="view.spia.row.total">Total received (period)</td><td>${money0(r.total_received_usd)}</td></tr>
                    <tr class="emph"><td data-i18n="view.spia.row.interest">Interest credited</td>
                        <td class="${r.interest_earned_usd >= 0 ? 'pos' : 'neg'}">${money0(r.interest_earned_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#spia-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: assumed rate 0 -> 10%; y: payout years 5 -> 35. Output: monthly income.
    const xVals = enh.linspace(0, 10, 5);
    const yVals = enh.linspace(5, 35, 5);
    const { cells } = await enh.runSensitivity({ base, xKey: 'annual_rate_pct', yKey: 'payout_years', xVals, yVals, compute: (b) => api.calcSpia(b), pick: (r) => (r ? r.monthly_income_usd : null) });
    if (!viewIsCurrent(tok)) return;
    panel.innerHTML = enh.renderSensitivityTable({ xVals, yVals, cells, fmt: (v) => (v == null ? '—' : '$' + Math.round(v)), xfmt: (v) => v.toFixed(1) + '%', yfmt: (v) => v.toFixed(0) + 'y', xName: t('view.spia.label.annual_rate_pct') || 'Rate', yName: t('view.spia.label.payout_years') || 'Years' });
}

File: crates/traderview-core/src/spia.rs
//! SPIA — single-premium immediate annuity payout.
//!
//! Hand an insurer a lump sum and get a guaranteed income stream starting
//! now. The monthly payment is the annuity that exhausts the premium over
//! the payout period at the insurer's assumed rate — the same present-value
//! annuity math as a loan payment, run in reverse:
//!
//!   * monthly = premium × i / (1 − (1 + i)^−n), i = rate/12, n = months
//!   * payout rate = annual income / premium (what insurers quote)
//!   * total received = monthly × n (premium + the interest credited)
//!
//! This is the period-certain core; a true life annuity prices in mortality,
//! but the payout-over-life-expectancy approximation is the deterministic
//! heart of it. Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct SpiaInput {
    pub premium_usd: f64,
    /// Expected payout period in years (e.g. life expectancy − current age).
    pub payout_years: f64,
    /// Insurer's assumed/credited annual rate.
    pub annual_rate_pct: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct SpiaResult {
    pub monthly_income_usd: f64,
    pub annual_income_usd: f64,
    /// Annual income as a percent of the premium — the quoted payout rate.
    pub payout_rate_pct: f64,
    /// Nominal total received over the payout period.
    pub total_received_usd: f64,
    /// Total received − premium (the interest credited over the period).
    pub interest_earned_usd: f64,
}

pub fn analyze(i: &SpiaInput) -> SpiaResult {
    let premium = i.premium_usd.max(0.0);
    let n = (i.payout_years * 12.0).max(0.0);
    let monthly_rate = i.annual_rate_pct / 100.0 / 12.0;

    let monthly = if n <= 0.0 {
        0.0
    } else if monthly_rate.abs() < 1e-12 {
        premium / n
    } else {
        premium * monthly_rate / (1.0 - (1.0 + monthly_rate).powf(-n))
    };

    let annual = monthly * 12.0;
    let total = monthly * n;

    SpiaResult {
        monthly_income_usd: monthly,
        annual_income_usd: annual,
        payout_rate_pct: if premium > 0.0 { annual / premium * 100.0 } else { 0.0 },
        total_received_usd: total,
        interest_earned_usd: total - premium,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn inp(premium: f64, years: f64, rate: f64) -> SpiaInput {
        SpiaInput { premium_usd: premium, payout_years: years, annual_rate_pct: rate }
    }

    #[test]
    fn zero_rate_is_straight_line() {
        // 100k over 20y = 240 months → 416.67/mo.
        let r = analyze(&inp(100_000.0, 20.0, 0.0));
        assert!((r.monthly_income_usd - 100_000.0 / 240.0).abs() < 1e-6);
    }

    #[test]
    fn with_rate_uses_annuity_formula() {
        // 100k, 20y, 5%: i=0.0041667, n=240 → ~659.96/mo.
        let r = analyze(&inp(100_000.0, 20.0, 5.0));
        let i: f64 = 0.05 / 12.0;
        let expected = 100_000.0 * i / (1.0 - (1.0 + i).powf(-240.0));
        assert!((r.monthly_income_usd - expected).abs() < 1e-6);
        assert!(r.monthly_income_usd > 600.0 && r.monthly_income_usd < 700.0);
    }

    #[test]
    fn annual_is_monthly_times_12() {
        let r = analyze(&inp(100_000.0, 20.0, 5.0));
        assert!((r.annual_income_usd - r.monthly_income_usd * 12.0).abs() < 1e-9);
    }

    #[test]
    fn payout_rate_is_annual_over_premium() {
        let r = analyze(&inp(100_000.0, 20.0, 5.0));
        assert!((r.payout_rate_pct - r.annual_income_usd / 100_000.0 * 100.0).abs() < 1e-9);
    }

    #[test]
    fn total_received_is_monthly_times_months() {
        let r = analyze(&inp(100_000.0, 20.0, 5.0));
        assert!((r.total_received_usd - r.monthly_income_usd * 240.0).abs() < 1e-6);
    }

    #[test]
    fn higher_rate_raises_monthly() {
        let low = analyze(&inp(100_000.0, 20.0, 3.0));
        let high = analyze(&inp(100_000.0, 20.0, 6.0));
        assert!(high.monthly_income_usd > low.monthly_income_usd);
    }

    #[test]
    fn longer_period_lowers_monthly() {
        let short = analyze(&inp(100_000.0, 10.0, 5.0));
        let long = analyze(&inp(100_000.0, 30.0, 5.0));
        assert!(long.monthly_income_usd < short.monthly_income_usd);
    }

    #[test]
    fn interest_earned_positive_with_rate() {
        // Zero-rate: total = premium, no interest.
        let flat = analyze(&inp(100_000.0, 20.0, 0.0));
        assert!(flat.interest_earned_usd.abs() < 1e-3);
        // With rate: total received exceeds the premium.
        let r = analyze(&inp(100_000.0, 20.0, 5.0));
        assert!(r.interest_earned_usd > 0.0);
    }
}
